Send report emails about test suites

ana opened the pull request:
This is a first approach to send emails with the result of the test suites. These reports will be improved after some testing, specially now that we should be having more results in the database. For the moment, they only allow to trigger emails for reporting all the tests ran for the same tree+branch+kernel, like this:

```
curl -X POST -H "Authorization: $EMAIL_AUTH_TOKEN" -H "Content-Type: application/json" -d '{"job": "'$TREE_NAME'", "kernel": "'$GIT_DESCRIBE'", "git_branch": "'$BRANCH'", "report_type": "test", "format": ["txt"], "send_to": ["ana@collabora.com"], "delay": 300}' ${API}/send
```

The emails generated have the following format:

```
Subject: Test suite result for local/master -  v4.17-6625-g1c8c5a9d38f6

Test suite result for local/master  - v4.17-6625-g1c8c5a9d38f6

Testsuite: lava

  *  Tree:        local
  *  Branch:      master
  *  URL:         /home/ana/KernelCI/src/linux
  *  Kernel:      v4.17-6625-g1c8c5a9d38f6
  *  Config:      defconfig
  *  Config Full: defconfig
  *  Lab Name:    lab-agl
  *  Board:       qemu
  *  Git Commit:  1c8c5a9d38f607c0b6fd12c91cbe1a4418762a21
  *  Date:        2018-06-11 00:07:32.196000

  * Test cases:

    * job : PASS
    * auto-login-action : PASS
    * http-download-ramdisk : PASS
    * http-download-kernel : PASS
    * validate : PASS

Testsuite: lava

  *  Tree:        local
  *  Branch:      master
  *  URL:         /home/ana/KernelCI/src/linux
  *  Kernel:      v4.17-6625-g1c8c5a9d38f6
  *  Config:      defconfig
  *  Config Full: defconfig
  *  Lab Name:    lab-agl
  *  Board:       qemu
  *  Git Commit:  1c8c5a9d38f607c0b6fd12c91cbe1a4418762a21
  *  Date:        2018-06-11 13:17:00.531000

  * Test cases:

    * validate : PASS
    * http-download-ramdisk : PASS
    * http-download-kernel : PASS
    * test-overlay-v4l2 : PASS
    * test-install-overlay-v4l2 : PASS
    * test-runscript-overlay-v4l2 : PASS
    * apply-overlay-guest : PASS
    * auto-login-action : PASS
    * 0_v4l2 : PASS
    * job : PASS

Testsuite: v4l2

  *  Tree:        local
  *  Branch:      master
  *  URL:         /home/ana/KernelCI/src/linux
  *  Kernel:      v4.17-6625-g1c8c5a9d38f6
  *  Config:      defconfig
  *  Config Full: defconfig
  *  Lab Name:    lab-agl
  *  Board:       qemu
  *  Git Commit:  1c8c5a9d38f607c0b6fd12c91cbe1a4418762a21
  *  Date:        2018-06-11 13:17:00.402000

  * Test cases:

    * v4l2-compliance : FAIL

```

khilman commented:
This looks like a good start.
Curious why your example has to sections for "Testsuite: lava"

Also, it would be nice to have a high-level summary at the Testsuite level.  e.g. something like
Testsuite: lava   N tests, Pass rate M% (X PASS, Y FAIL, Z SKIP)

ana commented:
@khilman 
> This looks like a good start.
> Curious why your example has to sections for "Testsuite: lava"
That was a mistake

ana commented:
I have updated the PR with the following changes:
- Remove the common data from every testsuite (tree, branch, kernel, git_url, git_commit), they're only shown once on the top of the mail now
- Do not show all the "lava" test suites, they're the build test and the kselftest tests. The kselftests should be updated to start appearing as "kselftest" test suites in the database.
- Add the stats lines with the number of total tests, passed, failed and skip tests

Example mail for  khilman/to-build - v4.18-rc8-53-gb4a2f821c4a1
https://gist.github.com/ana/af7f153384e95588132c588dac983215

And for stable/linux-4.14.y - v4.14.62
https://gist.github.com/ana/f5ba07193329234fdfd0f40a4f0f019c

ana commented:
hi! any comments or suggestion on this PR?
The reports are a bit long, but this is the first version. I'm planning to add some improvements to also allow sending the reports filtered by lab/board/test. Right now, I'd like to be able to start getting reports for the current test suites.

khilman commented:
IMO, Let's merge this and then iterate/improve.

Some thoughts:

It's missing a high-level summary at the top.  e.g. how many test suites ran?  how many passed/failed?  on what boards?

For example, something like for each test suite ran, how many sets/cases pass/fail/skip for each, and how many boards were each run on.  The idea is that from the first page of the email report, you have a pretty good idea whether most things are passing/failing without having to scroll into the details.


Suites    Sets   Cases   Boards
------------------------------------------------------
sleep:     n/a     X/Y/Z      3
usb
simple
cros-ec
igt
v4l2

gctucker commented on app/handlers/send.py:
```diff
@@ -11,7 +11,7 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-"""This module is used to send boot and build report email."""
+"""This module is used to send boot, build and test suite report emails."""
```
and also bisections... Maybe this could just say "to send report emails".

gctucker commented on app/taskqueue/tasks/report.py:
```diff
@@ -177,6 +178,41 @@ def send_bisect_report(report_data, email_opts, base_path=utils.BASE_PATH):
+    """
+    report_id = "-".join([job, git_branch, kernel])
+    utils.LOG.info("Sending test suite report email for '{}'".format(report_id))
+    status = "ERROR"
```
Not strictly required as this gets initialised when calling `utils.emails.send_email()`.

gctucker commented on app/utils/report/templates/test.html:
```diff
@@ -0,0 +1,63 @@
+            {% for tc in t.test_case_list %}
+            <tr><th>{{ tc.name }}:</th><td>{{ tc.status }}</td></tr>
+            {%- endfor %}
+            </table>
```
Small indentation glitch :)

gctucker commented on app/utils/report/templates/test.html:
```diff
@@ -0,0 +1,63 @@
+<!doctype html>
```
Do you have an example of what the HTML version of the report looks like?

gctucker commented on app/utils/report/test.py:
```diff
@@ -0,0 +1,149 @@
+# This program is free software: you can redistribute it and/or modify
```
We should also have a `utils.report.tests.test_test_report` unit test to cover this new module (and one for bisections as well I guess).

gctucker replied:
Also, does this pass pep8 and existing unit tests?

gctucker commented on app/utils/report/test.py:
```diff
@@ -0,0 +1,149 @@
+        ts['total_fail'] = len([e for e in resultlist if e=="FAIL"])
+        ts['total_skip'] = len([e for e in resultlist if e=="SKIP"])
+
+    # Remove all the "lava" test suites, they're the boot tests
```
It's not quite accurate that the `lava` test suites are the boot tests.  Rather, boot tests only have a `lava` test suite.  Ultimately it would be good to deal with failing boot results for any test plan (say, report that a `usb` test plan failed to complete) as that would also include the plain `boot` test plan without artificially creating a test suite for it.

For the comment here, I think it's better to say that the `lava` test suite doesn't contain any actual results from the defined test plan.

ana commented:
I just pushed v3 addressing the remarks from @gctucker. For the unit test part, the coverage is the same than master and adding `utils.report.tests.test_test_report` is still pending.
I also made some improvements in the templates, currently the emails looks like:
 *  txt: https://gist.github.com/ana/a02b85909985c5c21fe75be83826a9c8
*  html: https://gist.github.com/ana/d17fd6d5609b6616a166e33ae16573aa (you'll need to safe the raw gist as mail.html and read it with your browser)

ana commented:
@khilman @gctucker  the HTML was mangled by thunderbird, I have updated the gist and if you save as a html now it should look good :)
About the colors, none of the HTML reports right now have any, if somebody wants to write a nice CSS we can colour of all of them at the same time :)

gctucker commented:
Thanks, looks good to me.

khilman commented:
Looks good now, thanks.
